doc/vmk.1
.\" $Id: vmk.man,v 1.14 2025/04/17 07:48:27 lhf Exp $
.TH VMK 1 "$Date: 2025/04/15 07:11:27 $"
.SH NAME
vmk \- vmk interpreter
.SH SYNOPSIS
.B vmk
[
.I options
]
[
.I script
[
.I args
]
]
.SH DESCRIPTION
.B vmk
is the standalone vmk interpreter.
It loads and executes vmk programs,
either in textual source form or
in precompiled binary form.
(Precompiled binaries are output by
.BR vmkc ,
the vmk compiler.)
.B vmk
can be used as a batch interpreter and also interactively.
.LP
After handling the
.IR options ,
the vmk program in file
.I script
is loaded and executed.
The
.I args
are available to
.I script
as strings in a global table named
.B arg
and also as arguments to its main fn.
When called without arguments,
.B vmk
behaves as
.B "vmk \-v \-i"
if the standard input is a terminal,
and as
.B "vmk \-"
otherwise.
.LP
In interactive mode,
.B vmk
prompts the user,
reads lines from the standard input,
and executes them as they are read.
If the line contains an expression,
then the line is evavmkted and the result is printed.
If a line does not contain a complete statement,
then a secondary prompt is displayed and
lines are read until a complete statement is formed or
a syntax error is found.
.LP
Before handling command line options and scripts,
.B vmk
checks the contents of the environment variables
.B VMK_INIT_5_4
and
.BR VMK_INIT ,
in that order.
If the contents are of the form
.RI '@ filename ',
then
.I filename
is executed.
Otherwise, the contents are assumed to be a vmk statement and is executed.
When
.B VMK_INIT_5_4
is defined,
.B VMK_INIT
is ignored.
.SH OPTIONS
.TP
.BI \-e " stat"
execute statement
.IR stat .
.TP
.B \-i
enter interactive mode after executing
.IR script .
.TP
.BI \-l " mod"
require library
.I mod
into global
.IR mod .
.TP
.BI \-l " g=mod"
require library
.I mod
into global
.IR g .
.TP
.B \-v
show version information.
.TP
.B \-E
ignore environment variables.
.TP
.B \-W
turn warnings on.
.TP
.B \-\-
stop handling options.
.TP
.B \-
stop handling options and execute the standard input as a file.
.SH ENVIRONMENT VARIABLES
The following environment variables affect the execution of
.BR vmk .
When defined,
the version-specific variants take priority
and the version-neutral variants are ignored.
.TP
.B VMK_INIT, VMK_INIT_5_4
Code to be executed before command line options and scripts.
.TP
.B VMK_PATH, VMK_PATH_5_4
Initial value of package.path,
the path used by require to search for vmk loaders.
.TP
.B VMK_CPATH, VMK_CPATH_5_4
Initial value of package.cpath,
the path used by require to search for C loaders.
.SH EXIT STATUS
If a script calls os.exit,
then
.B vmk
exits with the given exit status.
Otherwise,
.B vmk
exits
with EXIT_SUCCESS (0 on POSIX systems) if there were no errors
and
with EXIT_FAILURE (1 on POSIX systems) if there were errors.
Errors raised in interactive mode do not cause exits.
.SH DIAGNOSTICS
Error messages should be self explanatory.
.SH "SEE ALSO"
.BR vmkc (1)
.br
The documentation at codeberg.org/nginrsw/vmk,
especially section 7 of the reference manual.
.SH AUTHORS
Gillar Ajie Prasatya
.\" EOF


src/vmk.c
/*
** $Id: vmk.c $
** Vmk stand-alone interpreter
** See Copyright Notice in vmk.h
*/

#define vmk_c

#include "lprefix.h"


#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <signal.h>

#include "vmk.h"

#include "lauxlib.h"
#include "vmklib.h"


#if !defined(VMK_PROGNAME)
#define VMK_PROGNAME		"vmk"
#endif

#if !defined(VMK_INIT_VAR)
#define VMK_INIT_VAR		"VMK_INIT"
#endif

#define VMK_INITVARVERSION	VMK_INIT_VAR VMK_VERSUFFIX


static vmk_State *globalL = NULL;

static const char *progname = VMK_PROGNAME;


#if defined(VMK_USE_POSIX)   /* { */

/*
** Use 'sigaction' when available.
*/
static void setsignal (int sig, void (*handler)(int)) {
  struct sigaction sa;
  sa.sa_handler = handler;
  sa.sa_flags = 0;
  sigemptyset(&sa.sa_mask);  /* do not mask any signal */
  sigaction(sig, &sa, NULL);
}

#else           /* }{ */

#define setsignal            signal

#endif                               /* } */


/*
** Hook set by signal function to stop the interpreter.
*/
static void lstop (vmk_State *L, vmk_Debug *ar) {
  (void)ar;  /* unused arg. */
  vmk_sethook(L, NULL, 0, 0);  /* reset hook */
  vmkL_error(L, "interrupted!");
}


/*
** Function to be called at a C signal. Because a C signal cannot
** just change a Vmk state (as there is no proper synchronization),
** this function only sets a hook that, when called, will stop the
** interpreter.
*/
static void laction (int i) {
  int flag = VMK_MASKCALL | VMK_MASKRET | VMK_MASKLINE | VMK_MASKCOUNT;
  setsignal(i, SIG_DFL); /* if another SIGINT happens, terminate process */
  vmk_sethook(globalL, lstop, flag, 1);
}


static void print_usage (const char *badoption) {
  vmk_writestringerror("%s: ", progname);
  if (badoption[1] == 'e' || badoption[1] == 'l')
    vmk_writestringerror("'%s' needs argument\n", badoption);
  else
    vmk_writestringerror("unrecognized option '%s'\n", badoption);
  vmk_writestringerror(
  "usage: %s [options] [script [args]]\n"
  "Available options are:\n"
  "  -e stat   execute string 'stat'\n"
  "  -i        enter interactive mode after executing 'script'\n"
  "  -l mod    require library 'mod' into global 'mod'\n"
  "  -l g=mod  require library 'mod' into global 'g'\n"
  "  -v        show version information\n"
  "  -E        ignore environment variables\n"
  "  -W        turn warnings on\n"
  "  --        stop handling options\n"
  "  -         stop handling options and execute stdin\n"
  ,
  progname);
}


/*
** Prints an error message, adding the program name in front of it
** (if present)
*/
static void l_message (const char *pname, const char *msg) {
  if (pname) vmk_writestringerror("%s: ", pname);
  vmk_writestringerror("%s\n", msg);
}


/*
** Check whether 'status' is not OK and, if so, prints the error
** message on the top of the stack.
*/
static int report (vmk_State *L, int status) {
  if (status != VMK_OK) {
    const char *msg = vmk_tostring(L, -1);
    if (msg == NULL)
      msg = "(error message not a string)";
    l_message(progname, msg);
    vmk_pop(L, 1);  /* remove message */
  }
  return status;
}


/*
** Message handler used to run all chunks
*/
static int msghandler (vmk_State *L) {
  const char *msg = vmk_tostring(L, 1);
  if (msg == NULL) {  /* is error object not a string? */
    if (vmkL_callmeta(L, 1, "__tostring") &&  /* does it have a metamethod */
        vmk_type(L, -1) == VMK_TSTRING)  /* that produces a string? */
      return 1;  /* that is the message */
    else
      msg = vmk_pushfstring(L, "(error object is a %s value)",
                               vmkL_typename(L, 1));
  }
  vmkL_traceback(L, L, msg, 1);  /* append a standard traceback */
  return 1;  /* return the traceback */
}


/*
** Interface to 'vmk_pcall', which sets appropriate message function
** and C-signal handler. Used to run all chunks.
*/
static int docall (vmk_State *L, int narg, int nres) {
  int status;
  int base = vmk_gettop(L) - narg;  /* function index */
  vmk_pushcfunction(L, msghandler);  /* push message handler */
  vmk_insert(L, base);  /* put it under function and args */
  globalL = L;  /* to be available to 'laction' */
  setsignal(SIGINT, laction);  /* set C-signal handler */
  status = vmk_pcall(L, narg, nres, base);
  setsignal(SIGINT, SIG_DFL); /* reset C-signal handler */
  vmk_remove(L, base);  /* remove message handler from the stack */
  return status;
}


static void print_version (void) {
  vmk_writestring(VMK_COPYRIGHT, strlen(VMK_COPYRIGHT));
  vmk_writeline();
}


/*
** Create the 'arg' table, which stores all arguments from the
** command line ('argv'). It should be aligned so that, at index 0,
** it has 'argv[script]', which is the script name. The arguments
** to the script (everything after 'script') go to positive indices;
** other arguments (before the script name) go to negative indices.
** If there is no script name, assume interpreter's name as base.
** (If there is no interpreter's name either, 'script' is -1, so
** table sizes are zero.)
*/
static void createargtable (vmk_State *L, char **argv, int argc, int script) {
  int i, narg;
  narg = argc - (script + 1);  /* number of positive indices */
  vmk_createtable(L, narg, script + 1);
  for (i = 0; i < argc; i++) {
    vmk_pushstring(L, argv[i]);
    vmk_rawseti(L, -2, i - script);
  }
  vmk_setglobal(L, "arg");
}


static int dochunk (vmk_State *L, int status) {
  if (status == VMK_OK) status = docall(L, 0, 0);
  return report(L, status);
}


static int dofile (vmk_State *L, const char *name) {
  return dochunk(L, vmkL_loadfile(L, name));
}


static int dostring (vmk_State *L, const char *s, const char *name) {
  return dochunk(L, vmkL_loadbuffer(L, s, strlen(s), name));
}


/*
** Receives 'globname[=modname]' and runs 'globname = require(modname)'.
** If there is no explicit modname and globname contains a '-', cut
** the suffix after '-' (the "version") to make the global name.
*/
static int dolibrary (vmk_State *L, char *globname) {
  int status;
  char *suffix = NULL;
  char *modname = strchr(globname, '=');
  if (modname == NULL) {  /* no explicit name? */
    modname = globname;  /* module name is equal to global name */
    suffix = strchr(modname, *VMK_IGMARK);  /* look for a suffix mark */
  }
  else {
    *modname = '\0';  /* global name ends here */
    modname++;  /* module name starts after the '=' */
  }
  vmk_getglobal(L, "require");
  vmk_pushstring(L, modname);
  status = docall(L, 1, 1);  /* call 'require(modname)' */
  if (status == VMK_OK) {
    if (suffix != NULL)  /* is there a suffix mark? */
      *suffix = '\0';  /* remove suffix from global name */
    vmk_setglobal(L, globname);  /* globname = require(modname) */
  }
  return report(L, status);
}


/*
** Push on the stack the contents of table 'arg' from 1 to #arg
*/
static int pushargs (vmk_State *L) {
  int i, n;
  if (vmk_getglobal(L, "arg") != VMK_TTABLE)
    vmkL_error(L, "'arg' is not a table");
  n = (int)vmkL_len(L, -1);
  vmkL_checkstack(L, n + 3, "too many arguments to script");
  for (i = 1; i <= n; i++)
    vmk_rawgeti(L, -i, i);
  vmk_remove(L, -i);  /* remove table from the stack */
  return n;
}


static int handle_script (vmk_State *L, char **argv) {
  int status;
  const char *fname = argv[0];
  if (strcmp(fname, "-") == 0 && strcmp(argv[-1], "--") != 0)
    fname = NULL;  /* stdin */
  status = vmkL_loadfile(L, fname);
  if (status == VMK_OK) {
    int n = pushargs(L);  /* push arguments to script */
    status = docall(L, n, VMK_MULTRET);
  }
  return report(L, status);
}


/* bits of various argument indicators in 'args' */
#define has_error	1	/* bad option */
#define has_i		2	/* -i */
#define has_v		4	/* -v */
#define has_e		8	/* -e */
#define has_E		16	/* -E */


/*
** Traverses all arguments from 'argv', returning a mask with those
** needed before running any Vmk code or an error code if it finds any
** invalid argument. In case of error, 'first' is the index of the bad
** argument.  Otherwise, 'first' is -1 if there is no program name,
** 0 if there is no script name, or the index of the script name.
*/
static int collectargs (char **argv, int *first) {
  int args = 0;
  int i;
  if (argv[0] != NULL) {  /* is there a program name? */
    if (argv[0][0])  /* not empty? */
      progname = argv[0];  /* save it */
  }
  else {  /* no program name */
    *first = -1;
    return 0;
  }
  for (i = 1; argv[i] != NULL; i++) {  /* handle arguments */
    *first = i;
    if (argv[i][0] != '-')  /* not an option? */
        return args;  /* stop handling options */
    switch (argv[i][1]) {  /* else check option */
      case '-':  /* '--' */
        if (argv[i][2] != '\0')  /* extra characters after '--'? */
          return has_error;  /* invalid option */
        *first = i + 1;
        return args;
      case '\0':  /* '-' */
        return args;  /* script "name" is '-' */
      case 'E':
        if (argv[i][2] != '\0')  /* extra characters? */
          return has_error;  /* invalid option */
        args |= has_E;
        break;
      case 'W':
        if (argv[i][2] != '\0')  /* extra characters? */
          return has_error;  /* invalid option */
        break;
      case 'i':
        args |= has_i;  /* (-i implies -v) *//* FALLTHROUGH */
      case 'v':
        if (argv[i][2] != '\0')  /* extra characters? */
          return has_error;  /* invalid option */
        args |= has_v;
        break;
      case 'e':
        args |= has_e;  /* FALLTHROUGH */
      case 'l':  /* both options need an argument */
        if (argv[i][2] == '\0') {  /* no concatenated argument? */
          i++;  /* try next 'argv' */
          if (argv[i] == NULL || argv[i][0] == '-')
            return has_error;  /* no next argument or it is another option */
        }
        break;
      default:  /* invalid option */
        return has_error;
    }
  }
  *first = 0;  /* no script name */
  return args;
}


/*
** Processes options 'e' and 'l', which involve running Vmk code, and
** 'W', which also affects the state.
** Returns 0 if some code raises an error.
*/
static int runargs (vmk_State *L, char **argv, int n) {
  int i;
  for (i = 1; i < n; i++) {
    int option = argv[i][1];
    vmk_assert(argv[i][0] == '-');  /* already checked */
    switch (option) {
      case 'e':  case 'l': {
        int status;
        char *extra = argv[i] + 2;  /* both options need an argument */
        if (*extra == '\0') extra = argv[++i];
        vmk_assert(extra != NULL);
        status = (option == 'e')
                 ? dostring(L, extra, "=(command line)")
                 : dolibrary(L, extra);
        if (status != VMK_OK) return 0;
        break;
      }
      case 'W':
        vmk_warning(L, "@on", 0);  /* warnings on */
        break;
    }
  }
  return 1;
}


static int handle_vmkinit (vmk_State *L) {
  const char *name = "=" VMK_INITVARVERSION;
  const char *init = getenv(name + 1);
  if (init == NULL) {
    name = "=" VMK_INIT_VAR;
    init = getenv(name + 1);  /* try alternative name */
  }
  if (init == NULL) return VMK_OK;
  else if (init[0] == '@')
    return dofile(L, init+1);
  else
    return dostring(L, init, name);
}


/*
** {==================================================================
** Read-Eval-Print Loop (REPL)
** ===================================================================
*/

#if !defined(VMK_PROMPT)
#define VMK_PROMPT		"> "
#define VMK_PROMPT2		">> "
#endif

#if !defined(VMK_MAXINPUT)
#define VMK_MAXINPUT		512
#endif


/*
** vmk_stdin_is_tty detects whether the standard input is a 'tty' (that
** is, whether we're running vmk interactively).
*/
#if !defined(vmk_stdin_is_tty)	/* { */

#if defined(VMK_USE_POSIX)	/* { */

#include <unistd.h>
#define vmk_stdin_is_tty()	isatty(0)

#elif defined(VMK_USE_WINDOWS)	/* }{ */

#include <io.h>
#include <windows.h>

#define vmk_stdin_is_tty()	_isatty(_fileno(stdin))

#else				/* }{ */

/* ISO C definition */
#define vmk_stdin_is_tty()	1  /* assume stdin is a tty */

#endif				/* } */

#endif				/* } */


/*
** vmk_readline defines how to show a prompt and then read a line from
** the standard input.
** vmk_saveline defines how to "save" a read line in a "history".
** vmk_freeline defines how to free a line read by vmk_readline.
*/
#if !defined(vmk_readline)	/* { */

#if defined(VMK_USE_READLINE)	/* { */

#include <readline/readline.h>
#include <readline/history.h>
#define vmk_initreadline(L)	((void)L, rl_readline_name="vmk")
#define vmk_readline(L,b,p)	((void)L, ((b)=readline(p)) != NULL)
#define vmk_saveline(L,line)	((void)L, add_history(line))
#define vmk_freeline(L,b)	((void)L, free(b))

#else				/* }{ */

#define vmk_initreadline(L)  ((void)L)
#define vmk_readline(L,b,p) \
        ((void)L, fputs(p, stdout), fflush(stdout),  /* show prompt */ \
        fgets(b, VMK_MAXINPUT, stdin) != NULL)  /* get line */
#define vmk_saveline(L,line)	{ (void)L; (void)line; }
#define vmk_freeline(L,b)	{ (void)L; (void)b; }

#endif				/* } */

#endif				/* } */


/*
** Return the string to be used as a prompt by the interpreter. Leave
** the string (or nil, if using the default value) on the stack, to keep
** it anchored.
*/
static const char *get_prompt (vmk_State *L, int firstline) {
  if (vmk_getglobal(L, firstline ? "_PROMPT" : "_PROMPT2") == VMK_TNIL)
    return (firstline ? VMK_PROMPT : VMK_PROMPT2);  /* use the default */
  else {  /* apply 'tostring' over the value */
    const char *p = vmkL_tolstring(L, -1, NULL);
    vmk_remove(L, -2);  /* remove original value */
    return p;
  }
}

/* mark in error messages for incomplete statements */
#define EOFMARK		"<eof>"
#define marklen		(sizeof(EOFMARK)/sizeof(char) - 1)


/*
** Check whether 'status' signals a syntax error and the error
** message at the top of the stack ends with the above mark for
** incomplete statements.
*/
static int incomplete (vmk_State *L, int status) {
  if (status == VMK_ERRSYNTAX) {
    size_t lmsg;
    const char *msg = vmk_tolstring(L, -1, &lmsg);
    if (lmsg >= marklen && strcmp(msg + lmsg - marklen, EOFMARK) == 0) {
      vmk_pop(L, 1);
      return 1;
    }
  }
  return 0;  /* else... */
}


/*
** Prompt the user, read a line, and push it into the Vmk stack.
*/
static int pushline (vmk_State *L, int firstline) {
  char buffer[VMK_MAXINPUT];
  char *b = buffer;
  size_t l;
  const char *prmt = get_prompt(L, firstline);
  int readstatus = vmk_readline(L, b, prmt);
  if (readstatus == 0)
    return 0;  /* no input (prompt will be popped by caller) */
  vmk_pop(L, 1);  /* remove prompt */
  l = strlen(b);
  if (l > 0 && b[l-1] == '\n')  /* line ends with newline? */
    b[--l] = '\0';  /* remove it */
  if (firstline && b[0] == '=')  /* for compatibility with 5.2, ... */
    vmk_pushfstring(L, "return %s", b + 1);  /* change '=' to 'return' */
  else
    vmk_pushlstring(L, b, l);
  vmk_freeline(L, b);
  return 1;
}


/*
** Try to compile line on the stack as 'return <line>;'; on return, stack
** has either compiled chunk or original line (if compilation failed).
*/
static int addreturn (vmk_State *L) {
  const char *line = vmk_tostring(L, -1);  /* original line */
  const char *retline = vmk_pushfstring(L, "return %s;", line);
  int status = vmkL_loadbuffer(L, retline, strlen(retline), "=stdin");
  if (status == VMK_OK) {
    vmk_remove(L, -2);  /* remove modified line */
    if (line[0] != '\0')  /* non empty? */
      vmk_saveline(L, line);  /* keep history */
  }
  else
    vmk_pop(L, 2);  /* pop result from 'vmkL_loadbuffer' and modified line */
  return status;
}


/*
** Read multiple lines until a complete Vmk statement
*/
static int multiline (vmk_State *L) {
  for (;;) {  /* repeat until gets a complete statement */
    size_t len;
    const char *line = vmk_tolstring(L, 1, &len);  /* get what it has */
    int status = vmkL_loadbuffer(L, line, len, "=stdin");  /* try it */
    if (!incomplete(L, status) || !pushline(L, 0)) {
      vmk_saveline(L, line);  /* keep history */
      return status;  /* cannot or should not try to add continuation line */
    }
    vmk_pushliteral(L, "\n");  /* add newline... */
    vmk_insert(L, -2);  /* ...between the two lines */
    vmk_concat(L, 3);  /* join them */
  }
}


/*
** Read a line and try to load (compile) it first as an expression (by
** adding "return " in front of it) and second as a statement. Return
** the final status of load/call with the resulting function (if any)
** in the top of the stack.
*/
static int loadline (vmk_State *L) {
  int status;
  vmk_settop(L, 0);
  if (!pushline(L, 1))
    return -1;  /* no input */
  if ((status = addreturn(L)) != VMK_OK)  /* 'return ...' did not work? */
    status = multiline(L);  /* try as command, maybe with continuation lines */
  vmk_remove(L, 1);  /* remove line from the stack */
  vmk_assert(vmk_gettop(L) == 1);
  return status;
}


/*
** Prints (calling the Vmk 'print' function) any values on the stack
*/
static void l_print (vmk_State *L) {
  int n = vmk_gettop(L);
  if (n > 0) {  /* any result to be printed? */
    vmkL_checkstack(L, VMK_MINSTACK, "too many results to print");
    vmk_getglobal(L, "print");
    vmk_insert(L, 1);
    if (vmk_pcall(L, n, 0, 0) != VMK_OK)
      l_message(progname, vmk_pushfstring(L, "error calling 'print' (%s)",
                                             vmk_tostring(L, -1)));
  }
}


/*
** Do the REPL: repeatedly read (load) a line, evavmkte (call) it, and
** print any results.
*/
static void doREPL (vmk_State *L) {
  int status;
  const char *oldprogname = progname;
  progname = NULL;  /* no 'progname' on errors in interactive mode */
  vmk_initreadline(L);
  while ((status = loadline(L)) != -1) {
    if (status == VMK_OK)
      status = docall(L, 0, VMK_MULTRET);
    if (status == VMK_OK) l_print(L);
    else report(L, status);
  }
  vmk_settop(L, 0);  /* clear stack */
  vmk_writeline();
  progname = oldprogname;
}

/* }================================================================== */


/*
** Main body of stand-alone interpreter (to be called in protected mode).
** Reads the options and handles them all.
*/
static int pmain (vmk_State *L) {
  int argc = (int)vmk_tointeger(L, 1);
  char **argv = (char **)vmk_touserdata(L, 2);
  int script;
  int args = collectargs(argv, &script);
  int optlim = (script > 0) ? script : argc; /* first argv not an option */
  vmkL_checkversion(L);  /* check that interpreter has correct version */
  if (args == has_error) {  /* bad arg? */
    print_usage(argv[script]);  /* 'script' has index of bad arg. */
    return 0;
  }
  if (args & has_v)  /* option '-v'? */
    print_version();
  if (args & has_E) {  /* option '-E'? */
    vmk_pushboolean(L, 1);  /* signal for libraries to ignore env. vars. */
    vmk_setfield(L, VMK_REGISTRYINDEX, "VMK_NOENV");
  }
  vmkL_openlibs(L);  /* open standard libraries */
  createargtable(L, argv, argc, script);  /* create table 'arg' */
  vmk_gc(L, VMK_GCRESTART);  /* start GC... */
  vmk_gc(L, VMK_GCGEN, 0, 0);  /* ...in generational mode */
  if (!(args & has_E)) {  /* no option '-E'? */
    if (handle_vmkinit(L) != VMK_OK)  /* run VMK_INIT */
      return 0;  /* error running VMK_INIT */
  }
  if (!runargs(L, argv, optlim))  /* execute arguments -e and -l */
    return 0;  /* something failed */
  if (script > 0) {  /* execute main script (if there is one) */
    if (handle_script(L, argv + script) != VMK_OK)
      return 0;  /* interrupt in case of error */
  }
  if (args & has_i)  /* -i option? */
    doREPL(L);  /* do read-eval-print loop */
  else if (script < 1 && !(args & (has_e | has_v))) { /* no active option? */
    if (vmk_stdin_is_tty()) {  /* running in interactive mode? */
      print_version();
      doREPL(L);  /* do read-eval-print loop */
    }
    else dofile(L, NULL);  /* executes stdin as a file */
  }
  vmk_pushboolean(L, 1);  /* signal no errors */
  return 1;
}


int main (int argc, char **argv) {
  int status, result;
  vmk_State *L = vmkL_newstate();  /* create state */
  if (L == NULL) {
    l_message(argv[0], "cannot create state: not enough memory");
    return EXIT_FAILURE;
  }
  vmk_gc(L, VMK_GCSTOP);  /* stop GC while building state */
  vmk_pushcfunction(L, &pmain);  /* to call 'pmain' in protected mode */
  vmk_pushinteger(L, argc);  /* 1st argument */
  vmk_pushlightuserdata(L, argv); /* 2nd argument */
  status = vmk_pcall(L, 2, 1, 0);  /* do the call */
  result = vmk_toboolean(L, -1);  /* get result */
  report(L, status);
  vmk_close(L);
  return (result && status == VMK_OK) ? EXIT_SUCCESS : EXIT_FAILURE;
}

